<a href="https://colab.research.google.com/github/ferneyreina19/IA-Aplicacion-a-la-Economia-1/blob/main/Practica_limpieza_Dairon_Ferney_Reina_Sabogal.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Inteligencia Artificial con Aplicaciones en Economía I**

- 👩‍🏫 **Profesora:** [Lina María Castro](https://www.linkedin.com/in/lina-maria-castro)  
- 📧 **Email:** [lmcastroco@gmail.com](mailto:lmcastroco@gmail.com)  
- 🎓 **Universidad:** Universidad Externado de Colombia - Facultad de Economía

# **Parte práctica parcial: Preparación y limpieza de datos**🧹

**IMPORTANTE**: Guarda una copia de este notebook en tu Google Drive o computador.

**Nombre estudiante:** Dairon Ferney Reina Sabogal

**Forma de entrega**

Jupyter Notebook publicado en su cuenta de Github con el nombre “Practica_limpieza_nombre_apellido_estudiante.ipynb”. Máximo a las 9 a.m., debes enviar link del notebook al correo lina.castro6@uexternado.edu.co, de lo contrario, no será tenido en cuenta.

**Instrucciones Generales**

- Puedes utilizar tus apuntes y notebooks y entrar al repositorio del curso. Sin embargo, **NO puedes utilizar IA para desarrollar tus respuestas**. **En caso de descubrir que estás usando IA** para el desarrollo de los ejercicios, **tu nota** de este segundo parcial **será cero**.
- Completa el código en las celdas marcadas con `### TU CÓDIGO AQUÍ ###`. Puedes añadir más celdas si lo requieres.
- Después de cada punto, ejecuta la celda de `assert` que se proporciona. Si no aparece ningún error, ¡tu respuesta es correcta!
- No modifiques las celdas de `assert` ni las celdas de Markdown.

**Objetivos de Aprendizaje**

1. **Aplicar técnicas de importación y consolidación de múltiples fuentes de datos** utilizando la librería `pandas`.
2.  **Ejecutar un proceso completo de limpieza de datos**, incluyendo el manejo de valores nulos, duplicados y la estandarización de tipos de datos.
3.  **Preparar un conjunto de datos real (GEIH del DANE)** para un análisis económico, sentando las bases para futuros modelos de machine learning.

# Ejercicio 1: Importar y unir datos (20%)

El DANE publica los microdatos de la GEIH mensualmente. Tu primera tarea es cargar los datos de los primeros seis meses de 2025 y armar una sola base.

**Los archivos necesarios los encuentras en el repositorio del curso, carpeta "datasets", carpeta "geih_2025".**

1. Usa un ciclo "for" para abrir los archivos CSV, ten en cuenta que estas bases usan como separador ";" y encoding='latin1'. La variable 'DPTO' y 'RAMA4D_R4' debes cargarlas como string.

In [ ]:
import numpy as np
import pandas as pd
import os as os
import warnings
warnings.filterwarnings('ignore')

In [ ]:
### TU CÓDIGO AQUÍ ###

pd.options.display.max_columns = None
# En los dataframes, mostrar los float con dos decimales
pd.options.display.float_format = '{:,.2f}'.format

path = '/content'
os.chdir(path)

df_2025 = [ f for f in os.listdir(path) if f.endswith('.csv')]

lista_dfs = []
for files in df_2025:
  try:
    df = pd.read_csv(files, sep=';',
                     encoding='latin1',
                     dtype={'DPTO': str, 'RAMA4D_R4': str})
    lista_dfs.append(df)
    print(f"Archivo {files} cargado correctamente.")
  except Exception as e:
      print(f"Error al cargar el archivo {files}")


Archivo df_jun.csv cargado correctamente.
Archivo df_ene.csv cargado correctamente.
Archivo df_feb.csv cargado correctamente.
Archivo df_may.csv cargado correctamente.
Archivo df_mar.csv cargado correctamente.
Archivo df_abr.csv cargado correctamente.


2. Une todos los dataframes para obtener una base de la GEIH con toda la información de enero a junio de 2025. El nuevo dataframe debe llamarse `df_geih`

In [ ]:
### TU CÓDIGO AQUÍ ###
df_geih = pd.concat(lista_dfs)
df_geih.head()

PERIODO  MES   PER  DIRECTORIO  SECUENCIA_P  ORDEN  HOGAR  REGIS  AREA  \
0  20250626    6  2025     8238974            1      1      1     10   NaN   
1  20250626    6  2025     8238974            1      2      1     10   NaN   
2  20250626    6  2025     8238975            1      1      1     10   NaN   
3  20250626    6  2025     8238975            1      2      1     10   NaN   
4  20250626    6  2025     8238975            1      3      1     10   NaN   

   CLASE  FEX_C18 DPTO  PT  P6016  P3271  P6040  P6030S1  P6030S3  P6050  \
0      2    67.48   08   1      1      2     66     9.00 1,958.00      1   
1      2    67.48   08   1      2      1     32     6.00 1,993.00      3   
2      2    54.34   08   1      1      1     76     1.00 1,949.00      1   
3      2    54.34   08   1      2      1     49    11.00 1,975.00      9   
4      2    54.34   08   1      3      2     52    10.00 1,972.00     13   

   P6083  P6083S1  P6081  P6081S1  P2057  P2059  P2061  P6080 P6080S1  \
0      3      NaN      3      NaN      1    NaN      1      6     NaN   
1      1     1.00      2      NaN      1    NaN      1      6     NaN   
2      3      NaN      3      NaN      1    NaN      1      6     NaN   
3      3      NaN      3      NaN      1    NaN      1      6     NaN   
4      2      NaN      2      NaN      1    NaN      1      6     NaN   

   P6080S1A1  P6070  P6071  P6071S1  P6090  P6100  P6110  P6120  P1906S1  \
0        NaN   4.00    NaN      NaN      1   3.00    NaN    NaN        4   
1        NaN   6.00    NaN      NaN      1   3.00    NaN    NaN        4   
2        NaN   5.00    NaN      NaN      1   3.00    NaN    NaN        4   
3        NaN   2.00   1.00     3.00      1   3.00    NaN    NaN        4   
4        NaN   2.00   1.00     2.00      1   3.00    NaN    NaN        4   

   P1906S2  P1906S3  P1906S4  P1906S5  P1906S6  P1906S7  P1906S8  P6160  \
0        4        3        4        4        4        4        4   1.00   
1        4        4        4        4        4        4        4   1.00   
2        4        4        4        4        4        4        4   1.00   
3        4        4        4        4        4        4        4   2.00   
4        4        4        4        4        4        4        4   2.00   

   P6170  P3041  P3042  P3042S1  P3042S2  P3043  P3043S1  P3038  P3039  \
0   2.00    NaN   5.00     2.00      NaN   2.00      NaN   1.00   2.00   
1   2.00    NaN   5.00     1.00      NaN   1.00      NaN   2.00   1.00   
2   2.00    NaN   3.00     3.00      NaN    NaN      NaN   2.00   1.00   
3   2.00    NaN   1.00     0.00      NaN    NaN      NaN   2.00   1.00   
4   2.00    NaN   1.00     0.00      NaN    NaN      NaN   1.00   2.00   

   POB_MAY18   FT  P3044S2  P6440  P6450  P6460  P6460S1  P6400  P6410  P6422  \
0       1.00 1.00   113.00   2.00    NaN    NaN      NaN    NaN    NaN    NaN   
1       1.00 1.00   113.00   2.00    NaN    NaN      NaN    NaN    NaN    NaN   
2       1.00 1.00   113.00   2.00    NaN    NaN      NaN    NaN    NaN    NaN   
3       1.00  NaN      NaN    NaN    NaN    NaN      NaN    NaN    NaN    NaN   
4       1.00 1.00   145.00   2.00    NaN    NaN      NaN    NaN    NaN    NaN   

   P6420S2  P6424S1  P6424S2  P6424S3  P6424S5  P6426  P6430 P6430S1  P3045S1  \
0      NaN      NaN      NaN      NaN      NaN  84.00   4.00     NaN      NaN   
1      NaN      NaN      NaN      NaN      NaN  84.00   6.00     NaN      NaN   
2      NaN      NaN      NaN      NaN      NaN 660.00   4.00     NaN      NaN   
3      NaN      NaN      NaN      NaN      NaN    NaN    NaN     NaN      NaN   
4      NaN      NaN      NaN      NaN      NaN   2.00   4.00     NaN      NaN   

   P3045S2  P3045S3  P3046  P3363  P9440  P6500  P3364  P3364S1  P6510  \
0      NaN      NaN    NaN    NaN    NaN    NaN    NaN      NaN    NaN   
1      NaN      NaN    NaN    NaN    NaN    NaN    NaN      NaN    NaN   
2      NaN      NaN    NaN    NaN    NaN    NaN    NaN      NaN    NaN   
3      NaN      

3. Reinicia el índice de `df_geih`. Asegúrate de que el índice antiguo no se añada como una nueva columna.

In [ ]:
### TU CÓDIGO AQUÍ ###
df_geih = pd.concat(lista_dfs, ignore_index=True)
df_geih.head()

PERIODO  MES   PER  DIRECTORIO  SECUENCIA_P  ORDEN  HOGAR  REGIS  AREA  \
0  20250626    6  2025     8238974            1      1      1     10   NaN   
1  20250626    6  2025     8238974            1      2      1     10   NaN   
2  20250626    6  2025     8238975            1      1      1     10   NaN   
3  20250626    6  2025     8238975            1      2      1     10   NaN   
4  20250626    6  2025     8238975            1      3      1     10   NaN   

   CLASE  FEX_C18 DPTO  PT  P6016  P3271  P6040  P6030S1  P6030S3  P6050  \
0      2    67.48   08   1      1      2     66     9.00 1,958.00      1   
1      2    67.48   08   1      2      1     32     6.00 1,993.00      3   
2      2    54.34   08   1      1      1     76     1.00 1,949.00      1   
3      2    54.34   08   1      2      1     49    11.00 1,975.00      9   
4      2    54.34   08   1      3      2     52    10.00 1,972.00     13   

   P6083  P6083S1  P6081  P6081S1  P2057  P2059  P2061  P6080 P6080S1  \
0      3      NaN      3      NaN      1    NaN      1      6     NaN   
1      1     1.00      2      NaN      1    NaN      1      6     NaN   
2      3      NaN      3      NaN      1    NaN      1      6     NaN   
3      3      NaN      3      NaN      1    NaN      1      6     NaN   
4      2      NaN      2      NaN      1    NaN      1      6     NaN   

   P6080S1A1  P6070  P6071  P6071S1  P6090  P6100  P6110  P6120  P1906S1  \
0        NaN   4.00    NaN      NaN      1   3.00    NaN    NaN        4   
1        NaN   6.00    NaN      NaN      1   3.00    NaN    NaN        4   
2        NaN   5.00    NaN      NaN      1   3.00    NaN    NaN        4   
3        NaN   2.00   1.00     3.00      1   3.00    NaN    NaN        4   
4        NaN   2.00   1.00     2.00      1   3.00    NaN    NaN        4   

   P1906S2  P1906S3  P1906S4  P1906S5  P1906S6  P1906S7  P1906S8  P6160  \
0        4        3        4        4        4        4        4   1.00   
1        4        4        4        4        4        4        4   1.00   
2        4        4        4        4        4        4        4   1.00   
3        4        4        4        4        4        4        4   2.00   
4        4        4        4        4        4        4        4   2.00   

   P6170  P3041  P3042  P3042S1  P3042S2  P3043  P3043S1  P3038  P3039  \
0   2.00    NaN   5.00     2.00      NaN   2.00      NaN   1.00   2.00   
1   2.00    NaN   5.00     1.00      NaN   1.00      NaN   2.00   1.00   
2   2.00    NaN   3.00     3.00      NaN    NaN      NaN   2.00   1.00   
3   2.00    NaN   1.00     0.00      NaN    NaN      NaN   2.00   1.00   
4   2.00    NaN   1.00     0.00      NaN    NaN      NaN   1.00   2.00   

   POB_MAY18   FT  P3044S2  P6440  P6450  P6460  P6460S1  P6400  P6410  P6422  \
0       1.00 1.00   113.00   2.00    NaN    NaN      NaN    NaN    NaN    NaN   
1       1.00 1.00   113.00   2.00    NaN    NaN      NaN    NaN    NaN    NaN   
2       1.00 1.00   113.00   2.00    NaN    NaN      NaN    NaN    NaN    NaN   
3       1.00  NaN      NaN    NaN    NaN    NaN      NaN    NaN    NaN    NaN   
4       1.00 1.00   145.00   2.00    NaN    NaN      NaN    NaN    NaN    NaN   

   P6420S2  P6424S1  P6424S2  P6424S3  P6424S5  P6426  P6430 P6430S1  P3045S1  \
0      NaN      NaN      NaN      NaN      NaN  84.00   4.00     NaN      NaN   
1      NaN      NaN      NaN      NaN      NaN  84.00   6.00     NaN      NaN   
2      NaN      NaN      NaN      NaN      NaN 660.00   4.00     NaN      NaN   
3      NaN      NaN      NaN      NaN      NaN    NaN    NaN     NaN      NaN   
4      NaN      NaN      NaN      NaN      NaN   2.00   4.00     NaN      NaN   

   P3045S2  P3045S3  P3046  P3363  P9440  P6500  P3364  P3364S1  P6510  \
0      NaN      NaN    NaN    NaN    NaN    NaN    NaN      NaN    NaN   
1      NaN      NaN    NaN    NaN    NaN    NaN    NaN      NaN    NaN   
2      NaN      NaN    NaN    NaN    NaN    NaN    NaN      NaN    NaN   
3      NaN      

In [ ]:
# Celda de verificación (no modificar)
assert df_geih.shape[0] == 408423, f"Error en las filas. Se esperaban 408423, pero se encontraron {df_geih.shape[0]}."
assert df_geih.shape[1] == 245, f"Error en las columnas. Se esperaban 245, pero se encontraron {df_geih.shape[1]}."
print("¡Las dimensiones del DataFrame son correctas! ✅")
assert 'index' not in df_geih.columns, "El índice antiguo no debe ser una columna."
assert df_geih.index.is_unique, "El índice debe ser único después de reiniciarlo."
print(f"¡Punto 1 completado con éxito! ✅ El DataFrame consolidado tiene {df_geih.shape[0]} filas y {df_geih.shape[1]} columnas.")

¡Las dimensiones del DataFrame son correctas! ✅
¡Punto 1 completado con éxito! ✅ El DataFrame consolidado tiene 408423 filas y 245 columnas.


# Ejercicio 2:  Seleccionar y renombrar columnas (5%)

La GEIH contiene cientos de variables. Para nuestro análisis del mercado laboral, nos enfocaremos en un subconjunto.

Adicionalmente, los nombres de las columnas en la GEIH son códigos. Para hacer el análisis más intuitivo, vamos a renombrarlas.

1. Crea un nuevo DataFrame llamado `df_empleo` que contenga únicamente las columnas que están en la lista `columnas_interes`.

In [ ]:
columnas_interes = ['DIRECTORIO', 'SECUENCIA_P', 'ORDEN', 'MES', 'CLASE', 'DPTO', 'FEX_C18', 'PT', 'P3271', 'P6040', 'P6080', 'P3043', 'OCI', 'P6430', 'RAMA4D_R4', 'INGLABO']

In [ ]:
### TU CÓDIGO AQUÍ ###
df_empleo = df_geih[columnas_interes]
df_empleo.head()

DIRECTORIO  SECUENCIA_P  ORDEN  MES  CLASE DPTO  FEX_C18  PT  P3271  P6040  \
0     8238974            1      1    6      2   08    67.48   1      2     66   
1     8238974            1      2    6      2   08    67.48   1      1     32   
2     8238975            1      1    6      2   08    54.34   1      1     76   
3     8238975            1      2    6      2   08    54.34   1      1     49   
4     8238975            1      3    6      2   08    54.34   1      2     52   

   P6080  P3043  OCI  P6430 RAMA4D_R4      INGLABO  
0      6   2.00 1.00   4.00      0113   950,000.00  
1      6   1.00 1.00   6.00      0113          NaN  
2      6    NaN 1.00   4.00      0113   700,000.00  
3      6    NaN  NaN    NaN       NaN          NaN  
4      6    NaN 1.00   4.00      0145 2,000,000.00

In [ ]:
# Celda de verificación (no modificar)
assert len(df_empleo.columns) == 16, "El número de columnas es incorrecto."
assert all(col in df_empleo.columns for col in columnas_interes), "Faltan columnas o los nombres son incorrectos."
print("¡Selección de columnas completada con éxito! ✅")
print("Dimensiones del nuevo DataFrame:", df_empleo.shape)

¡Selección de columnas completada con éxito! ✅
Dimensiones del nuevo DataFrame: (408423, 16)


2. Renombra las columnas del DataFrame `df_empleo` utilizando el diccionario `mapa_nombres`.

In [ ]:
mapa_nombres = {
'DIRECTORIO': 'id_vivienda',
'SECUENCIA_P': 'id_hogar',
'ORDEN': 'id_persona',
'MES': 'mes',
'CLASE': 'clase',
'DPTO': 'departamento',
'FEX_C18': 'factor_expansion',
'PT': 'poblacion_total',
'P3271': 'sexo',
'P6040': 'edad',
'P6080': 'grupo_etnico',
'P3043': 'nivel_educativo',
'OCI': 'ocupados',
'P6430': 'categoria_ocupacional',
'RAMA4D_R4': 'cod_ciiu',
'INGLABO': 'ingreso_laboral'}

In [ ]:
### TU CÓDIGO AQUÍ ###
df_empleo.rename(columns=mapa_nombres, inplace=True)
df_empleo.head()

id_vivienda  id_hogar  id_persona  mes  clase departamento  \
0      8238974         1           1    6      2           08   
1      8238974         1           2    6      2           08   
2      8238975         1           1    6      2           08   
3      8238975         1           2    6      2           08   
4      8238975         1           3    6      2           08   

   factor_expansion  poblacion_total  sexo  edad  grupo_etnico  \
0             67.48                1     2    66             6   
1             67.48                1     1    32             6   
2             54.34                1     1    76             6   
3             54.34                1     1    49             6   
4             54.34                1     2    52             6   

   nivel_educativo  ocupados  categoria_ocupacional cod_ciiu  ingreso_laboral  
0             2.00      1.00                   4.00     0113       950,000.00  
1             1.00      1.00                   6.00     0113              NaN  
2              NaN      1.00                   4.00     0113       700,000.00  
3              NaN       NaN                    NaN      NaN              NaN  
4              NaN      1.00                   4.00     0145     2,000,000.00

In [ ]:
# Celda de verificación (no modificar)
nombres_nuevos_esperados = list(mapa_nombres.values())
condicion_exitosa = all(col in df_empleo.columns for col in nombres_nuevos_esperados)
assert condicion_exitosa, "No todas las columnas esperadas se encuentran en el DataFrame. Revisa el diccionario o la operación de renombrado."
print("¡Verificación exitosa! Todas las columnas fueron renombradas correctamente. ✅")
print("¡Punto 2 completado con éxito! ✅")

¡Verificación exitosa! Todas las columnas fueron renombradas correctamente. ✅
¡Punto 2 completado con éxito! ✅


# Ejercicio 3: Revisión de duplicados (5%)

Un paso crucial en la limpieza es asegurar que no haya observaciones duplicadas que puedan sesgar nuestro análisis.

1. Verifica si hay filas duplicadas en `df_empleo`.

In [ ]:
### TU CÓDIGO AQUÍ ###
df_empleo[df_empleo.duplicated(keep=False)]

Empty DataFrame
Columns: [id_vivienda, id_hogar, id_persona, mes, clase, departamento, factor_expansion, poblacion_total, sexo, edad, grupo_etnico, nivel_educativo, ocupados, categoria_ocupacional, cod_ciiu, ingreso_laboral]
Index: []

2. Cuenta cuántas filas duplicadas existen.

In [ ]:
### TU CÓDIGO AQUÍ ###
df_empleo.duplicated().sum()
df_empleo.head()

id_vivienda  id_hogar  id_persona  mes  clase departamento  \
0      8238974         1           1    6      2           08   
1      8238974         1           2    6      2           08   
2      8238975         1           1    6      2           08   
3      8238975         1           2    6      2           08   
4      8238975         1           3    6      2           08   

   factor_expansion  poblacion_total  sexo  edad  grupo_etnico  \
0             67.48                1     2    66             6   
1             67.48                1     1    32             6   
2             54.34                1     1    76             6   
3             54.34                1     1    49             6   
4             54.34                1     2    52             6   

   nivel_educativo  ocupados  categoria_ocupacional cod_ciiu  ingreso_laboral  
0             2.00      1.00                   4.00     0113       950,000.00  
1             1.00      1.00                   6.00     0113              NaN  
2              NaN      1.00                   4.00     0113       700,000.00  
3              NaN       NaN                    NaN      NaN              NaN  
4              NaN      1.00                   4.00     0145     2,000,000.00

3. Elimina las filas duplicadas, si las hay.

In [ ]:
### TU CÓDIGO AQUÍ ###
df_empleo.drop_duplicates(inplace=True)

4. Verifica si hay filas duplicadas específicamente en las columnas 'id_vivienda', 'id_hogar', 'id_persona'.

In [ ]:
### TU CÓDIGO AQUÍ ###
df_empleo.duplicated(subset=['id_vivienda', 'id_hogar', 'id_persona']).sum()

np.int64(0)

In [ ]:
# Celda de verificación (no modificar)
assert df_empleo.duplicated().sum() == 0, "Aún existen filas duplicadas."
print("¡Punto 3 completado con éxito! ✅")

¡Punto 3 completado con éxito! ✅


# Ejercicio 4: Valores faltantes (5%)

Los datos del mundo real raramente están completos. Aquí, trataremos los valores faltantes en la columna de ingresos.


1. Calcula el número de valores nulos (`NaN`) en la columna `ingreso_laboral` **antes** de cualquier modificación.

In [ ]:
### TU CÓDIGO AQUÍ ###
df_empleo['ingreso_laboral'].isna()

0         False
1          True
2         False
3          True
4         False
          ...  
408418     True
408419     True
408420     True
408421    False
408422     True
Name: ingreso_laboral, Length: 408423, dtype: bool

2. En la GEIH, un ingreso faltante a menudo significa que la persona no reportó ingresos o no aplica (ej. no trabaja). Una estrategia común es imputar estos valores con `0`. Rellena los valores `NaN` en la columna `ingreso_laboral` con `0`.

In [ ]:
### TU CÓDIGO AQUÍ ###
df_empleo['ingreso_laboral'] = df_empleo['ingreso_laboral'].fillna(0)

In [ ]:
# Celda de verificación (no modificar)
assert df_empleo['ingreso_laboral'].isnull().sum() == 0, "La columna 'ingreso_laboral' todavía tiene valores nulos."
print("¡Punto 4 completado con éxito! ✅")

¡Punto 4 completado con éxito! ✅


# Ejercicio 5: Decodificar variables (20%)



Las variables categóricas a menudo necesitan ser "traducidas" para hacerlas más entendibles para el análisis.

1. Usa los siguientes diccionarios para reemplazar los códigos numéricos por etiquetas de texto en las siguientes columnas: 'clase', 'sexo', 'grupo_etnico', 'nivel_educativo' y 'categoria_ocupacional'. Muestra los conteos de las nuevas categorías usando `.value_counts()` para verificar el cambio en cada una de las columnas.

In [ ]:
dic_clase = {1: 'Cabeceras', 2: 'Centros poblados y rural disperso'}
dic_sexo = {1: 'Hombre', 2: 'Mujer'}
dic_etnico = {1: 'Indígena', 2: 'Gitano (Rom)', 3: 'Raizal del archipiélago', 4: 'Palenquero',
              5: 'Negro, mulato (afrodescendiente)', 6: 'Ninguno de los anteriores'}
dic_nivel_educativo = {1: 'Ninguno', 2: 'Educación media', 3: 'Educación media',
                       4: 'Educación media', 5: 'Técnica profesional', 6: 'Tecnológica',
                       7: 'Universitaria', 8: 'Especialización', 9: 'Maestría',
                       10: 'Doctorado', 99: 'No sabe, no informa'}
dic_cat_oc = {1: 'Asalariado', 2: 'Asalariado', 3: 'Asalariado',
               8: 'Asalariado', 4: 'Independiente', 5: 'Independiente',
               9: 'Independiente', 6: 'Trabajador sin remuneración',
               7: 'Trabajador sin remuneración'}

In [ ]:
df_empleo.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 408423 entries, 0 to 408422
Data columns (total 16 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   id_vivienda            408423 non-null  int64  
 1   id_hogar               408423 non-null  int64  
 2   id_persona             408423 non-null  int64  
 3   mes                    408423 non-null  int64  
 4   clase                  408423 non-null  int64  
 5   departamento           408423 non-null  object 
 6   factor_expansion       408423 non-null  float64
 7   poblacion_total        408423 non-null  int64  
 8   sexo                   408423 non-null  int64  
 9   edad                   408423 non-null  int64  
 10  grupo_etnico           408423 non-null  int64  
 11  nivel_educativo        208454 non-null  float64
 12  ocupados               176176 non-null  float64
 13  categoria_ocupacional  176176 non-null  float64
 14  cod_ciiu               176176 non-nu

In [ ]:
### TU CÓDIGO AQUÍ ###
df_empleo['clase']= df_empleo['clase'].map(dic_clase)
df_empleo['sexo']= df_empleo['sexo'].map(dic_sexo)
df_empleo['grupo_etnico']= df_empleo['grupo_etnico'].map(dic_etnico)
df_empleo['nivel_educativo']= df_empleo['nivel_educativo'].map(dic_nivel_educativo)
df_empleo['categoria_ocupacional']= df_empleo['categoria_ocupacional'].map(dic_cat_oc)


In [ ]:
# Celda de verificación (no modificar)
mapeos_a_verificar = {
    'clase': dic_clase,
    'sexo': dic_sexo,
    'grupo_etnico': dic_etnico,
    'nivel_educativo': dic_nivel_educativo,
    'categoria_ocupacional': dic_cat_ocu
}
for columna, diccionario in mapeos_a_verificar.items():
    es_numerica = pd.api.types.is_numeric_dtype(df_empleo[columna])
    assert not es_numerica, f"La columna '{columna}' todavía tiene un tipo de dato numérico. El reemplazo falló."
    valor_texto_ejemplo = list(diccionario.values())[0]
    categorias_unicas = df_empleo[columna].unique()
    assert valor_texto_ejemplo in categorias_unicas, f"No se encontró el valor '{valor_texto_ejemplo}' en la columna '{columna}'."
    valor_numerico_ejemplo = list(diccionario.keys())[0]
    assert valor_numerico_ejemplo not in categorias_unicas, f"El código numérico '{valor_numerico_ejemplo}' todavía existe en la columna '{columna}'."
print("¡Verificación completada con éxito! ✅ Todas las columnas categóricas fueron transformadas correctamente.")

¡Verificación completada con éxito! ✅ Todas las columnas categóricas fueron transformadas correctamente.


2. Utiliza el archivo "2025-01-14 Correlativa CIIU Rev4.xlsx", hoja 'CIIU 2022', para agregar al df_empleo las descripciones de los códigos CIIU a 4 dígitos. En df_empleo la columna con los códigos CIIU se llama 'cod_ciiu'. Si se agregan nuevas columnas al dataframe, distintas a la descripción del CIIU, debes borrarlas.

In [ ]:
### TU CÓDIGO AQUÍ ###
df_ciiu = pd.read_excel('/content/2025-01-14 Correlativa CIIU Rev4.xlsx', sheet_name='CIIU 2022',dtype={'RAMA4D_R4': str})
df_ciiu.value_counts()

RAMA4D_R4  DESCRIPCION_CIIU                                                     
9999       Actividad No Homologada a CIIU Rev 4                                     1
0000       Desconocido                                                              1
9499       Actividades de otras asociaciones n.c.p.                                 1
9492       Actividades de asociaciones políticas                                    1
9491       Actividades de asociaciones religiosas                                   1
                                                                                   ..
0111       Cultivo de cereales (excepto arroz), legumbres y semillas oleaginosas    1
0090       Actividad Personas Naturales                                             1
0082       Actividad Personas Naturales                                             1
0081       Actividad Personas Naturales                                             1
0020       Actividad Personas Naturales                                             1
Name: count, Length: 514, dtype: int64

In [ ]:
df_ciiu.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 514 entries, 0 to 513
Data columns (total 2 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   RAMA4D_R4         514 non-null    object
 1   DESCRIPCION_CIIU  514 non-null    object
dtypes: object(2)
memory usage: 8.2+ KB


In [ ]:
df_empleo = pd.merge(df_empleo, df_ciiu, left_on='cod_ciiu', right_on='RAMA4D_R4', how='left')
df_empleo.head()

id_vivienda  id_hogar  id_persona  mes                              clase  \
0      8238974         1           1    6  Centros poblados y rural disperso   
1      8238974         1           2    6  Centros poblados y rural disperso   
2      8238975         1           1    6  Centros poblados y rural disperso   
3      8238975         1           2    6  Centros poblados y rural disperso   
4      8238975         1           3    6  Centros poblados y rural disperso   

  departamento  factor_expansion  poblacion_total    sexo  edad  \
0           08             67.48                1   Mujer    66   
1           08             67.48                1  Hombre    32   
2           08             54.34                1  Hombre    76   
3           08             54.34                1  Hombre    49   
4           08             54.34                1   Mujer    52   

                grupo_etnico  nivel_educativo  ocupados  \
0  Ninguno de los anteriores  Educación media      1.00   
1  Ninguno de los anteriores          Ninguno      1.00   
2  Ninguno de los anteriores              NaN      1.00   
3  Ninguno de los anteriores              NaN       NaN   
4  Ninguno de los anteriores              NaN      1.00   

         categoria_ocupacional cod_ciiu  ingreso_laboral RAMA4D_R4  \
0                Independiente     0113       950,000.00      0113   
1  Trabajador sin remuneración     0113             0.00      0113   
2                Independiente     0113       700,000.00      0113   
3                          NaN      NaN             0.00       NaN   
4                Independiente     0145     2,000,000.00      0145   

                             DESCRIPCION_CIIU  
0  Cultivo de hortalizas, raíces y tubérculos  
1  Cultivo de hortalizas, raíces y tubérculos  
2  Cultivo de hortalizas, raíces y tubérculos  
3                                         NaN  
4                      Cría de aves de corral

In [ ]:
# Celda de verificación (no modificar)
assert 'DESCRIPCION_CIIU' in df_empleo.columns, \
    f"La columna '{'DESCRIPCION_CIIU'}' no se encontró. Revisa el nombre que le diste a la nueva columna."
assert df_empleo['DESCRIPCION_CIIU'].notna().sum() > 0, \
    f"La columna '{'DESCRIPCION_CIIU'}' está completamente vacía. Verifica las llaves del merge."
print("✅ Verificación del merge exitosa.")

✅ Verificación del merge exitosa.


3. Utiliza el archivo "2025-01-14 Divipola.xlsx", hoja 'Departamentos', para agregar al df_empleo los nombres de los departamentos. En df_empleo la columna con los códigos Divipola de los departamentos se llama 'departamento'. Si se agregan nuevas columnas al dataframe, distintas al nombre del departamento, debes borrarlas.

In [ ]:
### TU CÓDIGO AQUÍ ###
df_divipols = pd.read_excel('2025-01-14 DIVIPOLA.xlsx', sheet_name='Departamentos',dtype={'DPTO': str},skiprows=9 )
df_divipols.head()

Código        Nombre
0     05     ANTIOQUIA
1     08     ATLÁNTICO
2     11  BOGOTÁ, D.C.
3     13       BOLÍVAR
4     15        BOYACÁ

In [ ]:
df_empleo = pd.merge(df_empleo, df_divipols, left_on='departamento', right_on='Código', how='left')
df_empleo.columns

Index(['id_vivienda', 'id_hogar', 'id_persona', 'mes', 'clase', 'departamento',
       'factor_expansion', 'poblacion_total', 'sexo', 'edad', 'grupo_etnico',
       'nivel_educativo', 'ocupados', 'categoria_ocupacional', 'cod_ciiu',
       'ingreso_laboral', 'RAMA4D_R4', 'DESCRIPCION_CIIU', 'Código', 'Nombre'],
      dtype='object')

In [ ]:
Columns = ['id_vivienda', 'id_hogar', 'id_persona', 'mes', 'clase', 'departamento',
       'factor_expansion', 'poblacion_total', 'sexo', 'edad', 'grupo_etnico',
       'nivel_educativo', 'ocupados', 'categoria_ocupacional', 'cod_ciiu',
       'ingreso_laboral', 'DESCRIPCION_CIIU', 'Nombre']

df_empleo = df_empleo[Columns]
df_empleo.head()

id_vivienda  id_hogar  id_persona  mes                              clase  \
0      8238974         1           1    6  Centros poblados y rural disperso   
1      8238974         1           2    6  Centros poblados y rural disperso   
2      8238975         1           1    6  Centros poblados y rural disperso   
3      8238975         1           2    6  Centros poblados y rural disperso   
4      8238975         1           3    6  Centros poblados y rural disperso   

  departamento  factor_expansion  poblacion_total    sexo  edad  \
0           08             67.48                1   Mujer    66   
1           08             67.48                1  Hombre    32   
2           08             54.34                1  Hombre    76   
3           08             54.34                1  Hombre    49   
4           08             54.34                1   Mujer    52   

                grupo_etnico  nivel_educativo  ocupados  \
0  Ninguno de los anteriores  Educación media      1.00   
1  Ninguno de los anteriores          Ninguno      1.00   
2  Ninguno de los anteriores              NaN      1.00   
3  Ninguno de los anteriores              NaN       NaN   
4  Ninguno de los anteriores              NaN      1.00   

         categoria_ocupacional cod_ciiu  ingreso_laboral  \
0                Independiente     0113       950,000.00   
1  Trabajador sin remuneración     0113             0.00   
2                Independiente     0113       700,000.00   
3                          NaN      NaN             0.00   
4                Independiente     0145     2,000,000.00   

                             DESCRIPCION_CIIU     Nombre  
0  Cultivo de hortalizas, raíces y tubérculos  ATLÁNTICO  
1  Cultivo de hortalizas, raíces y tubérculos  ATLÁNTICO  
2  Cultivo de hortalizas, raíces y tubérculos  ATLÁNTICO  
3                                         NaN  ATLÁNTICO  
4                      Cría de aves de corral  ATLÁNTICO

In [ ]:
# Celda de verificación (no modificar)
assert 'Nombre' in df_empleo.columns, \
    f"La columna '{'Nombre'}' no se encontró. Revisa el nombre que le diste a la nueva columna."
assert df_empleo['Nombre'].notna().sum() > 0, \
    f"La columna '{'Nombre'}' está completamente vacía. Verifica las llaves del merge."
print("✅ Verificación del merge exitosa.")
assert df_empleo.shape[0] == 408423, f"Error en las filas. Se esperaban 408423, pero se encontraron {df_empleo.shape[0]}."
assert df_empleo.shape[1] == 18, f"Error en las columnas. Se esperaban 18, pero se encontraron {df_empleo.shape[1]}."
print("¡Las dimensiones del DataFrame son correctas! ✅")
print("\n¡Punto 5 completado con éxito! ✅")

✅ Verificación del merge exitosa.
¡Las dimensiones del DataFrame son correctas! ✅

¡Punto 5 completado con éxito! ✅


# Ejercicio 6: Crear nuevas variables y reorganizar columnas (10%)

La GEIH cuenta con un factor de expansión que permite extrapolar los datos de los encuestados a totales para la población colombiana. Usar este factor de expansión, permite calcular la población total del país, la población en edad de trabajar, el número de ocupados, el número de desempleados, entre otros datos.

1. Crea una nueva columna llamada "fex", la cual corresponderá a la columna 'factor_expansion' dividida en 6 (se divide en 6, ya que estamos tomando el dato de 6 meses).

In [ ]:
### TU CÓDIGO AQUÍ ###
df_empleo['fex'] = df_empleo['factor_expansion']/6

2. Crea una nueva columna llamada 'poblacion_total_ex', la cual corresponderá a la columna 'poblacion_total' multiplicada por la columna 'fex'.

In [ ]:
### TU CÓDIGO AQUÍ ###
df_empleo['poblacion_total_ex'] = df_empleo['poblacion_total'] * df_empleo['fex']

3. Crea una nueva columna llamada 'ocupados_ex', la cual corresponderá a la columna 'ocupados' multiplicada por la columna 'fex'.

In [ ]:
### TU CÓDIGO AQUÍ ###
df_empleo['ocupados_ex'] = df_empleo['ocupados'] * df_empleo['fex']
df_empleo[['fex', 'poblacion_total_ex', 'ocupados_ex']].head()

fex  poblacion_total_ex  ocupados_ex
0 11.25               11.25        11.25
1 11.25               11.25        11.25
2  9.06                9.06         9.06
3  9.06                9.06          NaN
4  9.06                9.06         9.06

4. Crea una nueva columna llamada 'joven', la cual debe decir 'Población joven' si la persona encuestada tiene entre 15 y 28 años (15 y 28 incluidos). De lo contrario, debe decir 'Resto de población'.

In [ ]:
### TU CÓDIGO AQUÍ ###
def joven(edad):
    if 15 <= edad <= 28:
        return 'Población joven'
    else:
        return 'Resto de población'

df_empleo['joven'] = df_empleo['edad'].apply(joven)
df_empleo[['fex', 'poblacion_total_ex', 'ocupados_ex', 'joven']].head()

fex  poblacion_total_ex  ocupados_ex               joven
0 11.25               11.25        11.25  Resto de población
1 11.25               11.25        11.25  Resto de población
2  9.06                9.06         9.06  Resto de población
3  9.06                9.06          NaN  Resto de población
4  9.06                9.06         9.06  Resto de población

5. Reorganiza las columnas con el siguiente orden: 'id_vivienda', 'id_hogar', 'id_persona', 'mes', 'clase', 'departamento', 'Nombre', 'factor_expansion', 'fex', 'poblacion_total', 'poblacion_total_ex', 'sexo', 'edad', 'joven', 'grupo_etnico', 'nivel_educativo', 'ocupados', 'ocupados_ex', 'categoria_ocupacional', 'cod_ciiu', 'DESCRIPCION_CIIU', 'ingreso_laboral'

In [ ]:
### TU CÓDIGO AQUÍ ###
organizacion = ['id_vivienda', 'id_hogar', 'id_persona',
                'mes', 'clase', 'departamento', 'Nombre',
                'factor_expansion', 'fex', 'poblacion_total',
                'poblacion_total_ex', 'sexo', 'edad', 'joven',
                'grupo_etnico', 'nivel_educativo', 'ocupados',
                'ocupados_ex', 'categoria_ocupacional', 'cod_ciiu',
                'DESCRIPCION_CIIU', 'ingreso_laboral']

df_empleo = df_empleo[organizacion]
df_empleo.head()

id_vivienda  id_hogar  id_persona  mes                              clase  \
0      8238974         1           1    6  Centros poblados y rural disperso   
1      8238974         1           2    6  Centros poblados y rural disperso   
2      8238975         1           1    6  Centros poblados y rural disperso   
3      8238975         1           2    6  Centros poblados y rural disperso   
4      8238975         1           3    6  Centros poblados y rural disperso   

  departamento     Nombre  factor_expansion   fex  poblacion_total  \
0           08  ATLÁNTICO             67.48 11.25                1   
1           08  ATLÁNTICO             67.48 11.25                1   
2           08  ATLÁNTICO             54.34  9.06                1   
3           08  ATLÁNTICO             54.34  9.06                1   
4           08  ATLÁNTICO             54.34  9.06                1   

   poblacion_total_ex    sexo  edad               joven  \
0               11.25   Mujer    66  Resto de población   
1               11.25  Hombre    32  Resto de población   
2                9.06  Hombre    76  Resto de población   
3                9.06  Hombre    49  Resto de población   
4                9.06   Mujer    52  Resto de población   

                grupo_etnico  nivel_educativo  ocupados  ocupados_ex  \
0  Ninguno de los anteriores  Educación media      1.00        11.25   
1  Ninguno de los anteriores          Ninguno      1.00        11.25   
2  Ninguno de los anteriores              NaN      1.00         9.06   
3  Ninguno de los anteriores              NaN       NaN          NaN   
4  Ninguno de los anteriores              NaN      1.00         9.06   

         categoria_ocupacional cod_ciiu  \
0                Independiente     0113   
1  Trabajador sin remuneración     0113   
2                Independiente     0113   
3                          NaN      NaN   
4                Independiente     0145   

                             DESCRIPCION_CIIU  ingreso_laboral  
0  Cultivo de hortalizas, raíces y tubérculos       950,000.00  
1  Cultivo de hortalizas, raíces y tubérculos             0.00  
2  Cultivo de hortalizas, raíces y tubérculos       700,000.00  
3                                         NaN             0.00  
4                      Cría de aves de corral     2,000,000.00

In [ ]:
# Celda de verificación (no modificar)
orden_esperado = [
    'id_vivienda', 'id_hogar', 'id_persona', 'mes', 'clase',
    'departamento', 'Nombre', 'factor_expansion', 'fex',
    'poblacion_total', 'poblacion_total_ex', 'sexo', 'edad', 'joven',
    'grupo_etnico', 'nivel_educativo', 'ocupados', 'ocupados_ex',
    'categoria_ocupacional', 'cod_ciiu', 'DESCRIPCION_CIIU', 'ingreso_laboral'
]
orden_actual = df_empleo.columns.tolist()
assert orden_actual == orden_esperado, \
    f"""El orden de las columnas es incorrecto.
    - Orden Esperado: {orden_esperado}
    - Orden Encontrado: {orden_actual}"""
assert df_empleo.shape[0] == 408423, f"Error en las filas. Se esperaban 408423, pero se encontraron {df_empleo.shape[0]}."
assert df_empleo.shape[1] == 22, f"Error en las columnas. Se esperaban 22, pero se encontraron {df_empleo.shape[1]}."
print("¡Las dimensiones del DataFrame son correctas! ✅")
print("\n¡Punto 6 completado con éxito! ✅")

¡Las dimensiones del DataFrame son correctas! ✅

¡Punto 6 completado con éxito! ✅


# Ejercicio 7: Cambiar tipo de dato (5%)



Para realizar cálculos y análisis correctamente, las columnas deben tener el tipo de dato adecuado (numérico, categórico, etc.).

1.  Muestra en pantalla los tipos de datos actuales de `df_empleo`.

In [ ]:
### TU CÓDIGO AQUÍ ###
df_empleo.dtypes

id_vivienda                int64
id_hogar                   int64
id_persona                 int64
mes                        int64
clase                     object
departamento              object
Nombre                    object
factor_expansion         float64
fex                      float64
poblacion_total            int64
poblacion_total_ex       float64
sexo                      object
edad                       int64
joven                     object
grupo_etnico              object
nivel_educativo           object
ocupados                 float64
ocupados_ex              float64
categoria_ocupacional     object
cod_ciiu                  object
DESCRIPCION_CIIU          object
ingreso_laboral          float64
dtype: object

2.  Las siguientes columnas representan categorías: 'clase', 'departamento', 'Nombre', 'sexo', 'joven', 'grupo_etnico', 'nivel_educativo', 'categoria_ocupacional', 'cod_ciiu', 'DESCRIPCION_CIIU'. Conviértelas al tipo `category` para optimizar el uso de memoria y facilitar análisis futuros.

In [ ]:
### TU CÓDIGO AQUÍ ###
categorias = ['clase', 'departamento',
            'Nombre', 'sexo', 'joven',
            'grupo_etnico', 'nivel_educativo',
            'categoria_ocupacional', 'cod_ciiu', 'DESCRIPCION_CIIU']
df_empleo[categorias] = df_empleo[categorias].astype('category')

df_empleo.dtypes

id_vivienda                 int64
id_hogar                    int64
id_persona                  int64
mes                         int64
clase                    category
departamento             category
Nombre                   category
factor_expansion          float64
fex                       float64
poblacion_total             int64
poblacion_total_ex        float64
sexo                     category
edad                        int64
joven                    category
grupo_etnico             category
nivel_educativo          category
ocupados                  float64
ocupados_ex               float64
categoria_ocupacional    category
cod_ciiu                 category
DESCRIPCION_CIIU         category
ingreso_laboral           float64
dtype: object

In [ ]:
# Celda de verificación (no modificar)
columnas_categoricas_esperadas = [
    'clase', 'departamento', 'Nombre', 'sexo', 'joven', 'grupo_etnico',
    'nivel_educativo', 'categoria_ocupacional', 'cod_ciiu',
    'DESCRIPCION_CIIU'
]
for columna in columnas_categoricas_esperadas:
    assert columna in df_empleo.columns, \
        f"La columna '{columna}' no fue encontrada en el DataFrame."
    es_categoria = pd.api.types.is_categorical_dtype(df_empleo[columna])
    assert es_categoria, \
        f"La columna '{columna}' no fue convertida al tipo 'category'. Su tipo actual es {df_empleo[columna].dtype}."
print("¡Verificación completada con éxito! ✅ Todas las columnas especificadas son del tipo 'category'.")
print("\n¡Punto 7 completado con éxito! ✅")

¡Verificación completada con éxito! ✅ Todas las columnas especificadas son del tipo 'category'.

¡Punto 7 completado con éxito! ✅


# Ejercicio 8: Ordenar el dataframe (5%)

1. Ordena el dataframe por la columna "edad" de mayor a menor. Observa la columna 'edad' y escribe en una celda de markdown cuál fue la edad mínima y la edad máxima de las personas que respondieron la encuesta de la GEIH.

In [ ]:
### TU CÓDIGO AQUÍ ###
df_empleo = df_empleo.sort_values(by='edad', ascending=False)
df_empleo.head()

id_vivienda  id_hogar  id_persona  mes  \
340200      8173637         1           2    4   
276696      8146261         1           3    3   
323169      8165410         1           3    3   
367040      8190663         1           1    4   
378399      8195350         1           1    4   

                                    clase departamento      Nombre  \
340200  Centros poblados y rural disperso           08   ATLÁNTICO   
276696                          Cabeceras           19       CAUCA   
323169                          Cabeceras           27       CHOCÓ   
367040                          Cabeceras           73      TOLIMA   
378399                          Cabeceras           44  LA GUAJIRA   

        factor_expansion   fex  poblacion_total  poblacion_total_ex   sexo  \
340200            190.48 31.75                1               31.75  Mujer   
276696            134.15 22.36                1               22.36  Mujer   
323169            482.95 80.49                1               80.49  Mujer   
367040            219.39 36.57                1               36.57  Mujer   
378399            568.77 94.80                1               94.80  Mujer   

        edad               joven                      grupo_etnico  \
340200   106  Resto de población         Ninguno de los anteriores   
276696   106  Resto de población         Ninguno de los anteriores   
323169   106  Resto de población  Negro, mulato (afrodescendiente)   
367040   105  Resto de población         Ninguno de los anteriores   
378399   105  Resto de población         Ninguno de los anteriores   

       nivel_educativo  ocupados  ocupados_ex categoria_ocupacional cod_ciiu  \
340200             NaN       NaN          NaN                   NaN      NaN   
276696             NaN       NaN          NaN                   NaN      NaN   
323169             NaN       NaN          NaN                   NaN      NaN   
367040             NaN       NaN          NaN                   NaN      NaN   
378399             NaN       NaN          NaN                   NaN      NaN   

       DESCRIPCION_CIIU  ingreso_laboral  
340200              NaN             0.00  
276696              NaN             0.00  
323169              NaN             0.00  
367040              NaN             0.00  
378399              NaN             0.00

#### **Escribe tu análisis aquí:**
##### Edad minima: 0 años
Edad maxima: 106 años


In [ ]:
# Celda de verificación (no modificar)
es_descendiente = df_empleo['edad'].is_monotonic_decreasing
assert es_descendiente, "La columna 'edad' no está ordenada correctamente de mayor a menor."
print("\n¡Punto 8 completado con éxito! ✅")


¡Punto 8 completado con éxito! ✅


# Ejercicio 9: Groupby (20%)

Agrupar los datos por categorías, nos permite realizar análisis más profundos y obtener insights valiosos, que a simple vista no vemos en la base de datos.

1. Crea un nuevo dataframe llamado "df_sexo" donde se agrupe por 'sexo' y se calcule la suma de las columnas 'poblacion_total_ex' y 'ocupados_ex'. Crea una nueva columna que se llame 'porcentaje_ocupados' donde dividas las columna 'ocupados_ex' sobre 'poblacion_total_ex'. Al final usa reset_index() para convertir las categorías de sexo en una columna.

In [ ]:
### TU CÓDIGO AQUÍ ###
df_sexo = df_empleo.groupby('sexo').agg(
    poblacion_total_ex=('poblacion_total_ex', 'sum'),
    ocupados_ex=('ocupados_ex', 'sum')
)
df_sexo['porcentaje_ocupados'] = (df_sexo['ocupados_ex'] / df_sexo['poblacion_total_ex'])*100
df_sexo = df_sexo.reset_index()
df_sexo

sexo  poblacion_total_ex   ocupados_ex  porcentaje_ocupados
0  Hombre       25,255,111.00 13,797,363.08                54.63
1   Mujer       26,682,244.83  9,707,731.65                36.38

In [ ]:
# Celda de verificación (no modificar)
import numpy as np
dimensiones_esperadas = (2, 4)
assert df_sexo.shape == dimensiones_esperadas, \
    f"Las dimensiones son incorrectas. Se esperaba {dimensiones_esperadas}, pero se encontró {df_sexo.shape}. ¿Aplicaste .reset_index()?"
assert 'sexo' in df_sexo.columns, \
    "La columna 'sexo' no se encontró. ¿Olvidaste usar .reset_index() al final?"
assert not isinstance(df_sexo.index, pd.Index) or df_sexo.index.name != 'sexo', \
    "El índice del DataFrame sigue siendo 'sexo'. Asegúrate de haber aplicado .reset_index()."
print("¡Verificación completada con éxito! ✅")

¡Verificación completada con éxito! ✅


2. Crea un nuevo dataframe llamado "df_educacion" donde se agrupe por 'nivel_educativo' y se calcule la suma de las columnas 'poblacion_total_ex' y 'ocupados_ex'. Crea una nueva columna que se llame 'porcentaje_ocupados' donde dividas las columna 'ocupados_ex' sobre 'poblacion_total_ex'. Al final usa reset_index() para convertir las categorías de nivel educativo en una columna.

In [ ]:
### TU CÓDIGO AQUÍ ###
df_educacion = df_empleo.groupby('nivel_educativo').agg(
    poblacion_total_ex=('poblacion_total_ex', 'sum'),
    ocupados_ex=('ocupados_ex', 'sum')
)
df_educacion['porcentaje_ocupados'] = (df_educacion['ocupados_ex'] / df_educacion['poblacion_total_ex'])*100
df_educacion = df_educacion.reset_index()
df_educacion

nivel_educativo  poblacion_total_ex  ocupados_ex  porcentaje_ocupados
0            Doctorado           46,649.18    39,950.47                85.64
1      Educación media       14,327,525.06 8,605,757.92                60.06
2      Especialización          965,236.30   777,759.47                80.58
3             Maestría          448,666.05   386,964.59                86.25
4              Ninguno        1,710,908.13   348,733.89                20.38
5  No sabe, no informa               94.78        37.60                39.68
6          Tecnológica        1,392,852.75 1,040,969.57                74.74
7  Técnica profesional        2,781,981.33 1,953,369.02                70.22
8        Universitaria        3,731,910.34 2,759,724.79                73.95

In [ ]:
# --- Celda de Verificación (No modificar) ---
import numpy as np
n_niveles_educativos = df_empleo['nivel_educativo'].nunique()
dimensiones_esperadas = (n_niveles_educativos, 4)
assert df_educacion.shape == dimensiones_esperadas, \
    f"Las dimensiones son incorrectas. Se esperaba {dimensiones_esperadas} (filas, columnas), pero se encontró {df_educacion.shape}."
columnas_esperadas = {'nivel_educativo', 'poblacion_total_ex', 'ocupados_ex', 'porcentaje_ocupados'}
assert set(df_educacion.columns) == columnas_esperadas, \
    f"Los nombres de las columnas no son correctos. Se esperaban {columnas_esperadas}."
assert 'nivel_educativo' in df_educacion.columns, \
    "La columna 'nivel_educativo' no se encontró. ¿Olvidaste usar .reset_index() al final?"
print("¡Verificación completada con éxito! ✅")

¡Verificación completada con éxito! ✅


3. En una celda de Markdown escribe qué conclusiones puedes sacar de la información obtenida en df_sexo y df_educacion.

#### **Escribe tu análisis aquí:**
##### **df_sexo**:  El porcentaje de hombres ocupados es del 54.63%, mientras que en las mujeres es del 36.38%. Esto indica una brecha de género considerable en el mercado laboral.

##### **df_educacion**: Se observa una brecha entre los niveles de educación superior. Aunque todos tienen una alta tasa de ocupación, hay una diferencia notable:

- Doctorado: 85.64%
- Maestría : 86.25%
- Universitaria: 73.95%
- Técnica profesional: 70.22%
- Tecnológica: 74.74%

Esto podría indicar que, aunque la formación en doctorado y maestria abarcan el porcentaje de ocupacion mas alto entre los niveles de educacion. La formacion universitaria tambien cuenta con un alto valor en el mercado, las carreras técnicas y tecnológicas también tienen una gran demanda y tasas de ocupación comparables.

In [ ]:
# Celda de verificación (no modificar)
print("\n¡Punto 9 completado con éxito! ✅")


¡Punto 9 completado con éxito! ✅


# Ejercicio 10: Exportar datos (5%)

Muchas veces requerimos exportar los resultados de nuestros análisis a Excel para compartir la información con otras personas.

Exporta los dataframes df_sexo y df_educacion a un Excel llamado 'Analisis_ocupados.xlsx'. Cada dataframe debe quedar en una hoja diferente. La hoja de 'df_sexo' se debe llamar 'sexo' y la hoja de 'df_educacion' se debe llamar 'educacion'.

In [ ]:
### TU CÓDIGO AQUÍ ###
with pd.ExcelWriter('Analisis_ocupados.xlsx') as writer:
    df_sexo.to_excel(writer, sheet_name='sexo', index=False)
    df_educacion.to_excel(writer, sheet_name='educacion', index=False)

In [ ]:
# Celda de verificación (no modificar)
import os
import pandas as pd
nombre_archivo = 'Analisis_ocupados.xlsx'
assert os.path.exists(nombre_archivo), \
    f"No se encontró el archivo '{nombre_archivo}'. ¿Lo exportaste con el nombre correcto?"
try:
    xls = pd.ExcelFile(nombre_archivo)
    nombres_hojas_encontrados = xls.sheet_names
    nombres_hojas_esperados = ['sexo', 'educacion']
    assert set(nombres_hojas_encontrados) == set(nombres_hojas_esperados), \
        f"Los nombres de las hojas no son correctos. Se esperaban {nombres_hojas_esperados} pero se encontraron {nombres_hojas_encontrados}."
    print("¡Verificación completada con éxito! ✅")
except Exception as e:
    print(f"Ocurrió un error al verificar el archivo: {e}")
print("\n¡Punto 10 completado con éxito! ✅")

¡Verificación completada con éxito! ✅

¡Punto 10 completado con éxito! ✅
